File: Liftr/StrengthWorkoutFinishCollapse.swift
import Foundation

enum StrengthWorkoutFinishCollapse {
    struct Line {
        let configId: Int
        let segmentsInRow: Int
        let reps: Int?
        let weightKg: Decimal?
        let rpe: Decimal?
        let restSec: Int?
        let weightSegments: [StrengthWeightSegWire]?
    }

    private struct CollapsedPersistRow {
        let count: Int
        let reps: Int?
        let weightKg: Decimal?
        let rpe: Decimal?
        let restSec: Int?
        let weightSegments: [StrengthWeightSegWire]?
    }

    private struct LegacyKey: Equatable {
        let reps: Int?
        let weightKg: Decimal?
        let rpe: Decimal?
        let restSec: Int?
    }

    static func buildExerciseSaveInputs(
        exerciseIds: [Int],
        performedByExercise: [Int: [Line]]
    ) -> [StrengthWorkoutExerciseSaveInput] {
        exerciseIds.map { exerciseId in
            let lines = performedByExercise[exerciseId] ?? []
            let chunks = chunkCompletedLines(lines)
            let rowsOut = chunks.flatMap { chunkToPersistRows($0) }
            let sets = rowsOut.map { row in
                StrengthWorkoutSetSaveInput(
                    set_number: row.count,
                    order_index: nil,
                    reps: row.reps,
                    weight_kg: row.weightKg.map { NSDecimalNumber(decimal: $0).doubleValue },
                    rpe: row.rpe.map { NSDecimalNumber(decimal: $0).doubleValue },
                    rest_sec: row.restSec,
                    weight_segments: row.weightSegments
                )
            }
            return StrengthWorkoutExerciseSaveInput(
                workout_exercise_id: exerciseId,
                exercise_id: nil,
                order_index: nil,
                notes: nil,
                custom_name: nil,
                sets: sets
            )
        }
    }

    private static func chunkCompletedLines(_ lines: [Line]) -> [[Line]] {
        guard !lines.isEmpty else { return [] }
        var out: [[Line]] = []
        var cur: [Line] = [lines[0]]
        for i in 1..<lines.count {
            let line = lines[i]
            if line.configId == cur[0].configId {
                cur.append(line)
            } else {
                out.append(cur)
                cur = [line]
            }
        }
        out.append(cur)
        return out
    }

    private static func collapseLegacyChunk(_ chunk: [Line]) -> [CollapsedPersistRow] {
        if chunk.isEmpty { return [] }
        var collapsed: [CollapsedPersistRow] = []
        var currentKey: LegacyKey?
        var count = 0
        for line in chunk {
            let key = LegacyKey(reps: line.reps, weightKg: line.weightKg, rpe: line.rpe, restSec: line.restSec)
            if let cur = currentKey, cur == key {
                count += 1
            } else {
                if let cur = currentKey {
                    collapsed.append(
                        CollapsedPersistRow(
                            count: count,
                            reps: cur.reps,
                            weightKg: cur.weightKg,
                            rpe: cur.rpe,
                            restSec: cur.restSec,
                            weightSegments: nil
                        )
                    )
                }
                currentKey = key
                count = 1
            }
        }
        if let cur = currentKey {
            collapsed.append(
                CollapsedPersistRow(
                    count: count,
                    reps: cur.reps,
                    weightKg: cur.weightKg,
                    rpe: cur.rpe,
                    restSec: cur.restSec,
                    weightSegments: nil
                )
            )
        }
        return collapsed
    }

    private static func collapseSegmentChunk(_ chunk: [Line]) -> [CollapsedPersistRow] {
        if chunk.count == 1, let f = chunk.first, let ws = f.weightSegments, ws.count >= 2 {
            let firstSeg = ws[0]
            return [
                CollapsedPersistRow(
                    count: 1,
                    reps: firstSeg.reps,
                    weightKg: Decimal(firstSeg.weight_kg),
                    rpe: f.rpe,
                    restSec: f.restSec,
                    weightSegments: ws
                )
            ]
        }
        let k = max(2, chunk.first?.segmentsInRow ?? 1)
        let n = chunk.count % k == 0 ? chunk.count / k : 1
        let tail = Array(chunk.suffix(k))
        let segs: [StrengthWeightSegWire] = tail.compactMap { line in
            guard let r = line.reps, let w = line.weightKg else { return nil }
            let d = NSDecimalNumber(decimal: w).doubleValue
            return StrengthWeightSegWire(reps: r, weight_kg: d)
        }
        if segs.count != tail.count {
            return [
                CollapsedPersistRow(
                    count: 1,
                    reps: tail.first?.reps,
                    weightKg: tail.first?.weightKg,
                    rpe: chunk.first(where: { $0.rpe != nil })?.rpe,
                    restSec: tail.last?.restSec,
                    weightSegments: nil
                )
            ]
        }
        let validSegs = segs.count >= 2 ? segs : nil
        return [
            CollapsedPersistRow(
                count: n,
                reps: tail.first?.reps,
                weightKg: tail.first?.weightKg,
                rpe: chunk.first(where: { $0.rpe != nil })?.rpe,
                restSec: tail.last?.restSec,
                weightSegments: validSegs
            )
        ]
    }

    private static func chunkToPersistRows(_ chunk: [Line]) -> [CollapsedPersistRow] {
        if chunk.isEmpty { return [] }
        if chunk[0].segmentsInRow <= 1 {
            return collapseLegacyChunk(chunk)
        }
        return collapseSegmentChunk(chunk)
    }
}

File: android/app/src/main/java/com/lilru/liftr/workout/StrengthWorkoutFinishCollapse.kt
package com.lilru.liftr.workout

import com.lilru.liftr.ui.active.CompletedSetLine
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

internal data class StrengthFinishSetPayload(
    val setNumber: Int,
    val reps: Int?,
    val weightKg: Double?,
    val rpe: Double?,
    val restSec: Int?,
    val weightSegments: JsonArray?
)

internal data class StrengthFinishExercisePayload(
    val workoutExerciseId: Int,
    val sets: List<StrengthFinishSetPayload>
)

internal object StrengthWorkoutFinishCollapse {
    private data class CollapsedPersistRow(
        val count: Int,
        val reps: Int?,
        val weightKg: Double?,
        val rpe: Double?,
        val restSec: Int?,
        val weightSegments: JsonArray? = null
    )

    private data class LegacyLineKey(val reps: Int?, val weightKg: Double?, val rpe: Double?, val restSec: Int?)
    private data class SegmentLineKey(
        val reps: Int?,
        val weightKg: Double?,
        val rpe: Double?,
        val restSec: Int?,
        val weightSegmentsRaw: String?
    )

    fun buildExercisePayloads(
        workoutExerciseIds: List<Int>,
        completedByExerciseId: Map<Int, List<CompletedSetLine>>
    ): List<StrengthFinishExercisePayload> {
        return workoutExerciseIds.map { weId ->
            val lines = completedByExerciseId[weId].orEmpty()
            val sets = chunkCompletedLines(lines)
                .flatMap { chunkToPersistRows(it) }
                .map { row ->
                    StrengthFinishSetPayload(
                        setNumber = row.count,
                        reps = row.reps,
                        weightKg = row.weightKg,
                        rpe = row.rpe,
                        restSec = row.restSec,
                        weightSegments = row.weightSegments
                    )
                }
            StrengthFinishExercisePayload(workoutExerciseId = weId, sets = sets)
        }
    }

    fun exerciseToJsonObject(exercise: StrengthFinishExercisePayload) = buildJsonObject {
        put("workout_exercise_id", exercise.workoutExerciseId)
        put(
            "sets",
            buildJsonArray {
                exercise.sets.forEach { set ->
                    add(
                        buildJsonObject {
                            put("set_number", set.setNumber)
                            set.reps?.let { put("reps", it) }
                            set.weightKg?.let { put("weight_kg", it) }
                            set.rpe?.let { put("rpe", it) }
                            set.restSec?.let { put("rest_sec", it) }
                            set.weightSegments?.let { put("weight_segments", it) }
                        }
                    )
                }
            }
        )
    }

    fun exercisesToJsonArray(exercises: List<StrengthFinishExercisePayload>) = buildJsonArray {
        exercises.forEach { add(exerciseToJsonObject(it)) }
    }

    private fun chunkCompletedLines(lines: List<CompletedSetLine>): List<List<CompletedSetLine>> {
        if (lines.isEmpty()) return emptyList()
        val out = mutableListOf<MutableList<CompletedSetLine>>()
        var cur = mutableListOf(lines.first())
        for (i in 1 until lines.size) {
            val line = lines[i]
            if (line.configId == cur.first().configId) {
                cur += line
            } else {
                out += cur
                cur = mutableListOf(line)
            }
        }
        out += cur
        return out
    }

    private fun collapseLegacyChunk(chunk: List<CompletedSetLine>): List<CollapsedPersistRow> {
        if (chunk.isEmpty()) return emptyList()
        val collapsed = mutableListOf<CollapsedPersistRow>()
        var currentKey: LegacyLineKey? = null
        var count = 0
        chunk.forEach { line ->
            val key = LegacyLineKey(line.reps, line.weightKg, line.rpe, line.restSec)
            if (currentKey != null && currentKey == key) {
                count += 1
            } else {
                if (currentKey != null) {
                    collapsed += CollapsedPersistRow(count, currentKey!!.reps, currentKey!!.weightKg, currentKey!!.rpe, currentKey!!.restSec, null)
                }
                currentKey = key
                count = 1
            }
        }
        if (currentKey != null) {
            collapsed += CollapsedPersistRow(count, currentKey!!.reps, currentKey!!.weightKg, currentKey!!.rpe, currentKey!!.restSec, null)
        }
        return collapsed
    }

    private fun collapseSegmentChunk(chunk: List<CompletedSetLine>): List<CollapsedPersistRow> {
        if (chunk.isEmpty()) return emptyList()
        val collapsed = mutableListOf<CollapsedPersistRow>()
        var currentKey: SegmentLineKey? = null
        var currentSegs: JsonArray? = null
        var count = 0
        fun flush() {
            val k = currentKey ?: return
            collapsed += CollapsedPersistRow(
                count = count,
                reps = k.reps,
                weightKg = k.weightKg,
                rpe = k.rpe,
                restSec = k.restSec,
                weightSegments = currentSegs
            )
        }
        chunk.forEach { line ->
            val key = SegmentLineKey(
                line.reps,
                line.weightKg,
                line.rpe,
                line.restSec,
                line.weightSegments?.toString()
            )
            if (currentKey != null && currentKey == key) {
                count += 1
            } else {
                if (currentKey != null) flush()
                currentKey = key
                currentSegs = line.weightSegments
                count = 1
            }
        }
        if (currentKey != null) flush()
        return collapsed
    }

    private fun chunkToPersistRows(chunk: List<CompletedSetLine>): List<CollapsedPersistRow> {
        if (chunk.isEmpty()) return emptyList()
        return if (chunk.first().segmentsInRow <= 1) {
            collapseLegacyChunk(chunk)
        } else {
            collapseSegmentChunk(chunk)
        }
    }
}
